src/validate.rs

```diff
@@ -1735,3 +1738,138 @@ mod test {
         assert_eq!(rrsig.verify_signed_data(&key, &signed_data), Ok(()));
     }
 }
+
+//------------ Nsec3HashError -------------------------------------------------
+
+/// An error when creating an NSEC3 hash.
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum Nsec3HashError {
+    /// The requested algorithm for NSEC3 hashing is not supported.
+    UnsupportedAlgorithm,
+
+    /// Data could not be appended to a buffer.
+    ///
+    /// This could indicate an out of memory condition.
+    AppendError,
+
+    /// The hashing process produced an invalid owner hash.
+    ///
+    /// See: [OwnerHashError](crate::rdata::nsec3::OwnerHashError)
+    OwnerHashError,
+
+    /// The hashing process produced a hash that already exists.
+    CollisionDetected,
+}
+
+///--- Display
+
+impl std::fmt::Display for Nsec3HashError {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        match self {
+            Nsec3HashError::UnsupportedAlgorithm => {
+                f.write_str("Unsupported algorithm")
+            }
+            Nsec3HashError::AppendError => {
+                f.write_str("Append error: out of memory?")
+            }
+            Nsec3HashError::OwnerHashError => {
+                f.write_str("Hashing produced an invalid owner hash")
+            }
+            Nsec3HashError::CollisionDetected => {
+                f.write_str("Hash collision detected")
+            }
+        }
+    }
+}
+
+/// Compute an [RFC 5155] NSEC3 hash using default settings.
+///
+/// See: [Nsec3param::default].
+///
+/// [RFC 5155]: https://www.rfc-editor.org/rfc/rfc5155
+pub fn nsec3_default_hash<N, HashOcts>(
+    owner: N,
+) -> Result<OwnerHash<HashOcts>, Nsec3HashError>
+where
+    N: ToName,
+    HashOcts: AsRef<[u8]> + EmptyBuilder + OctetsBuilder + Truncate,
+    for<'a> HashOcts: From<&'a [u8]>,
+{
+    let params = Nsec3param::<HashOcts>::default();
+    nsec3_hash(
+        owner,
+        params.hash_algorithm(),
+        params.iterations(),
+        params.salt(),
+    )
+}
+
+/// Compute an [RFC 5155] NSEC3 hash.
+///
+/// Computes an NSEC3 hash according to [RFC 5155] section 5:
+///
+/// > IH(salt, x, 0) = H(x || salt)
+/// > IH(salt, x, k) = H(IH(salt, x, k-1) || salt), if k > 0
+///
+/// Then the calculated hash of an owner name is:
+///
+/// > IH(salt, owner name, iterations),
+///
+/// Note that the `iterations` parameter is the number of _additional_
+/// iterations as defined in [RFC 5155] section 3.1.3.
+///
+/// [RFC 5155]: https://www.rfc-editor.org/rfc/rfc5155
+pub fn nsec3_hash<N, SaltOcts, HashOcts>(
```

Comment:
I think we need to avoid adding new stuff to src/validate.rs. Certainly not stuff that is needed for signing.